### PR TITLE
Модели и их сравнение

### PR DESCRIPTION
**Pull request by Gambitttttt**
Добавлена возможность обучения с помощью моделей DQN, Q_table и возможность сравнивать их с другими типами игроков

### CODE REVIEW

File: src/model.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
+import torch
+import numpy as np
+import torch.nn.functional as F
+from torch import nn, optim
+import os
+
+class DQN(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.fc1 = nn.Linear(4, 512)
```
Comment (joitandr) [inline src/model.py:10]: Я немного не в контексте, но вот такая матрица проекции кажется странной. Ты из 4 фичей создаёшь 512 фичей. Если это работает, то кажется должно работать и сильно меньшая проекция, например (4, 100)

File: src/model.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+        file_name = os.path.join(model_folder_path, file_name)
+        torch.save(self.state_dict(), file_name)
+
+    def load(self, file_name = 'model5000.pth'):
+        model_folder_path = './src'
+        file_name = os.path.join(model_folder_path, file_name)
+        self.load_state_dict(torch.load(file_name))
+        
+class trainer():
```
Comment (joitandr) [inline src/model.py:32]: классы обычно описывают в стиле [PascalCase](https://khalilstemmler.com/blogs/camel-case-snake-case-pascal-case/). То есть в твоём случае `class Trainer`

File: src/model.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    def __init__(self, model, lr, gamma):
```
Comment (joitandr) [inline src/model.py:33]: добавь аннотацию типов: `lr: float, gamma: float`
для model можно создать например базовый абстрактный класс с моделью и от него наследовать DQN и QTable, например так:
```python
class BaseQTableModel(nn.Module):
    def __init__(self, ...):
        super().__init__(...)
        ...
    
    def forward(self, ....):
        ...

class QTable(BaseQTableModel):
    ...
    
class DQN(BaseQTableModel):
    ...
    
class Trainer:
    def __init__(self, model: BaseQTableModel, lr: float, gamma: float):
        ... 
```

File: src/model.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        self.model = model
+        self.lr = lr
+        self.gamma = gamma
+        self.optimizer = optim.Adam(model.parameters(), lr = self.lr)
+        self.criterion = nn.MSELoss()
+
+    def train_step(self, state, action, reward, next_state, done):
```
Comment (joitandr) [inline src/model.py:40]: также аннотацию типов для state, action, reward, next_state и done. Поскольку у тебя это какие-то доменно специфичные типы будет - нужно описать что это такое

File: src/players.py [hunk 5]
```diff
@@ -40,27 +50,106 @@ def get_history(self,) -> t.List[Action]:
     def get_last_action(self) -> Action:
         return self.history[-1]
     
-    def action(self, num_options, money_available) -> Action:
+    def action(self, num_options, money_available, states, epsilon, n_games, decay, total_state, model) -> Action:
```
Comment (joitandr) [inline src/players.py:53]: очень запутанный код. Из того что я понял, ты пытаешься вызвать своё действие в зависимости от типа игрока - в таком случае можно сделать отдельные действия для каждого игрока и в основной функции вызывать одно какое-то действие, например так:

```python
ActionType = int # basically action type is field id
StateType = dict[int, dict[t.Any]] # player_id -> meta info of previous actions
PolicyType = t.Callable[[StateType], ActionType] # some function that takes in StateType and returns ActionType

def action_sber_lover(state: StateType, policy: PolicyType) -> ActionType:
   # do some sber lover stuff
    ...

def action_oil_lover(state: StateType, policy: PolicyType) -> ActionType:
    # do some oil_lover staff
    ...

def action_manufacturer(state: StateType, policy: PolicyType) -> ActionType:
    # do some manufacturer staff
    ...

playertype2action_function = {
    "manufacturer": action_manufacturer,
    "sber_lover": action_sber_lover,
    ...
}

def action(state: StateType, policy: PolicyType, player_type: str) -> ActionType:
    return playertype2action_function["player_type"](state, policy)
```

^ гораздо лаконичнее
